File: activemq-core/src/main/java/org/apache/activemq/transport/stomp/StompTransportFactory.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.activemq.transport.stomp;

import java.util.HashMap;
import java.util.Map;

import org.apache.activemq.broker.BrokerContext;
import org.apache.activemq.broker.BrokerService;
import org.apache.activemq.broker.BrokerServiceAware;
import org.apache.activemq.transport.MutexTransport;
import org.apache.activemq.transport.Transport;
import org.apache.activemq.transport.tcp.TcpTransportFactory;
import org.apache.activemq.util.IntrospectionSupport;
import org.apache.activemq.wireformat.WireFormat;

/**
 * A <a href="http://stomp.codehaus.org/">STOMP</a> transport factory
 */
public class StompTransportFactory extends TcpTransportFactory implements BrokerServiceAware {

    private BrokerContext brokerContext = null;

    protected String getDefaultWireFormatType() {
        return "stomp";
    }

    @SuppressWarnings("rawtypes")
    public Transport compositeConfigure(Transport transport, WireFormat format, Map options) {
        transport = new StompTransportFilter(transport, format, brokerContext);
        IntrospectionSupport.setProperties(transport, options);
        return super.compositeConfigure(transport, format, options);
    }

    public void setBrokerService(BrokerService brokerService) {
        this.brokerContext = brokerService.getBrokerContext();
    }

    @SuppressWarnings("rawtypes")
    @Override
    public Transport serverConfigure(Transport transport, WireFormat format, HashMap options) throws Exception {
        transport = super.serverConfigure(transport, format, options);

        MutexTransport mutex = transport.narrow(MutexTransport.class);
        if (mutex != null) {
            mutex.setSyncOnCommand(true);
        }

        return transport;
    }

    @Override
    protected Transport createInactivityMonitor(Transport transport, WireFormat format) {
        StompInactivityMonitor monitor = new StompInactivityMonitor(transport, format);

        StompTransportFilter filter = (StompTransportFilter) transport.narrow(StompTransportFilter.class);
        filter.setInactivityMonitor(monitor);

        return monitor;
    }
}

File: cms/activemqcms/src/activemq/transport/stomp/StompTransportFactory.cpp
/*
 * Copyright 2006 The Apache Software Foundation or its licensors, as
 * applicable.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "StompTransportFactory.h"
#include "StompTransport.h"
#include <sstream>

using namespace activemq::transport;
using namespace activemq::transport::stomp;
using namespace std;

////////////////////////////////////////////////////////////////////////////////
Transport* StompTransportFactory::createTransport( const char* brokerUrl ){
    
    string temp = brokerUrl;
    unsigned int ix = temp.find( ':' );
    if( ix == string::npos ){
        throw new ActiveMQException( "StompTransportFactory::createTransport - no port provided in url" );
    }
    
    brokerHost = temp.substr(0,ix);
    
    sscanf(brokerUrl+ix+1, "%d", &brokerPort );
    
    return new StompTransport( brokerHost.c_str(), brokerPort, "", "" );
}

////////////////////////////////////////////////////////////////////////////////        
Transport* StompTransportFactory::createTransport( const char* brokerUrl, 
    const char* userName,
    const char* password )
{
    this->userName = userName;
    this->password = password;
    
    parseUrl( brokerUrl );
    return new StompTransport( brokerHost.c_str(), 
        brokerPort, 
        userName, 
        password );
}

////////////////////////////////////////////////////////////////////////////////
void StompTransportFactory::parseUrl( const char* brokerUrl )
{   
    string url = brokerUrl;
    unsigned int portStartIx = url.rfind( ':' );
    if( portStartIx == string::npos || portStartIx==(url.length()-1) ){
        brokerPort = 61626;
    }
    else{
        
        try{
            stringstream stream( url.substr( portStartIx+1 ) );
            stream >> brokerPort;
        }catch( ... ){
            brokerPort = 61626;
            portStartIx = string::npos;
        }
    }
    
    brokerHost = url.substr( 0, portStartIx );    
}

